### PR TITLE
Move API Change Guidelines from wiki to docs/devel

### PR DESCRIPTION
**Pull request by MikeJeffrey**
https://github.com/GoogleCloudPlatform/kubernetes/issues/2633

### CODE REVIEW

**Comment by erictune** [comment]
Is it possible you forgot to `git rm docs/api-change-guidelines.md`

**Comment by MikeJeffrey** [comment]
I don't see docs/api-change-guidelines.md? Did I accidentally try to push it somewhere, or does it already exist? Sorry, always tripping up with git :)

**Comment by bgrant0607** [comment]
There is no docs/api-change-guidelines.md. This issue is about moving content from the wiki to git.

File: docs/devel/api-change-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# API Change Guidelines
+
+Freely change pkg/api/types.go.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: append:
```
, which is the internal system representation.
```

File: docs/devel/api-change-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# API Change Guidelines
+
+Freely change pkg/api/types.go. 
+
+Update the most recent versioned package's conversion functions. (In, e.g., pkg/api/v1beta1/conversion.go)
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: and defaulting pass (e.g., pkg/api/v1beta1/defaults.go)

File: docs/devel/api-change-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+If necessary, you may add fields to the versioned package's objects. You may not remove fields or rename them (or change the json/yaml tags), because that will break clients.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: or change default values

File: docs/devel/api-change-guidelines.md (hunk 2 above)
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Actually, all the older package's conversion functions are the ones that should be updated.

File: docs/devel/api-change-guidelines.md (hunk 3 above)
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Also:
All newly added fields must be documented using field description tags. We have a pre-submit hook to verify that descriptions exist for all API fields.

File: docs/devel/api-change-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+This allows us to make progress on the internal API without breaking clients. pkg/api/serialization_test.go should detect errors that will result in information being lost as a result of doing something wrong in modifying the API.
+
+At some point, we may want to cut a new version. Assume the new version is called v8 and the previous version is called v7:
+
+1. Copy what we have in pkg/api/types.go to pkg/api/v8/types.go.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: This should be:
1. `cp -r pkg/api/v7 pkg/api/v8`

File: docs/devel/api-change-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+2. Fix up links in pkg/api/v7's conversion package-- now they convert to pkg/api/v8 and not pkg/api.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Delete this line. It's no longer valid.

File: docs/devel/api-change-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+3. pkg/api/v8 gets a new register.go and a conversion.go, the latter having nothing to do yet.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Actually, delete points 3-5, also.

File: docs/devel/api-change-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+4. Update pkg/api/latest to use the new version as the default.
+5. Update RAML documentation. (We really need to start auto generating it...)
+
+(Note that this info may need to change when we finish cutting v1beta2.)
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Delete this line.

File: docs/devel/api-change-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+Q (Clayton): what are we going to do about old clients that PUT old API versions that lack a field in the new internal?  Let them clobber it?  Intelligently merge once we get the conflicting version from etcd?  Seems like we need to know what version of the API the client submitted a result in to know whether to merge a field or not.
```
Comment (bgrant0607) [inline docs/devel/api-change-guidelines.md]: Replace this comment with:

Clients, such as kubectl, that need to work generically with versioned API objects that may add new fields, multiple API versions, and API extensions should not decode/encode to/from the internal representation. They should work with the raw json/yaml. If the API schema is needed (e.g., for validation), it should be queried from the apiserver's Swagger endpoint.

Object generators, such as [`kubectl run-container`](kubectl-run-container.md), should target a specific version of the API, and should themselves be versioned.

**Comment by MikeJeffrey** [comment]
Messed this up with git.

**Pull request closed without merging**